### ISSUE TITLE
Can I register a callback to be called before or after GC?

### ISSUE THREAD

**Issue by noahfalk** [dotnet/coreclr#24104]
@ygc369 [Commented on another thread](https://github.com/dotnet/coreclr/pull/22866#issuecomment-484040178), but it is a new issue so moving it here:

>  @Maoni0
Can I register a callback to be called before or after GC? Just in my program, not in profile.

And the answer: You might be interested in [this technique](https://blogs.msdn.microsoft.com/maoni/2019/04/08/a-portable-way-to-get-gc-events-in-process-and-no-admin-privilege-with-10-lines-of-code-and-ability-to-dynamically-enable-disable-events/).

**Comment by noahfalk** [comment]
> @noahfalk Yes, that's what I want. But unfortunately, @Maoni0 didn't paste her complete code in that artical, so I still don't know how to use.

I don't understand, Maoni's post included [the full 101 line sample code](https://gist.github.com/Maoni0/e7e4c94aa0a23fb096ff09a5b84f9a48#file-fullgcevents-cs) which can be directly built and run without modification. I'm not sure what is lacking?

**Comment by ygc369** [comment]
@noahfalk Oh, I see, thank you.
I have the last question, can I know the GC generation when "GCStart" event happens?

**Comment by noahfalk** [comment]
Yeah you can, it is the 'Depth' property. There are [full docs](https://docs.microsoft.com/en-us/dotnet/framework/performance/garbage-collection-etw-events) on all the events.

You can print all the properties of all events with this example that enumerates the PayloadNames and Payload properties on the EventWrittenEventArgs.

```
    class MyEventListener : EventListener
    {
        protected override void OnEventSourceCreated(EventSource eventSource)
        {
            base.OnEventSourceCreated(eventSource);
            Console.WriteLine(eventSource.Name);
            if(eventSource.Name == "Microsoft-Windows-DotNETRuntime")
            {
                this.EnableEvents(eventSource, EventLevel.Informational, (EventKeywords)0x1);
            }
        }

        protected override void OnEventWritten(EventWrittenEventArgs eventData)
        {
            Console.WriteLine(eventData.EventName);
            for(int i = 0; i < eventData.PayloadNames.Count; i++)
            {
                Console.WriteLine("    " + eventData.PayloadNames[i] + ": " + eventData.Payload[i]);
            }
        }
    }
```

**Comment by ygc369** [comment]
@noahfalk 
Does "GCStart" event happen just before GC really starts to work?

**Comment by noahfalk** [comment]
No there is not that guarantee, EventListener is asynchronous so delivery occurs sometime after the underlying event.
